### PR TITLE
[demos][python] Add a new Python PaLM demo: Docs Agent

### PR DESCRIPTION
**Pull request by kyolee415**
Doc Agents allows PaLM API users to launch a chat application using their own documents as a dataset.

### CODE REVIEW

File: demos/palm/python/docs-agent/scripts/populate_vector_database.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+                                    print("Up to date content.")
+                                    unchanged_count += 1
+                                else:
+                                    collection.add(
+                                        documents=toFile,
+                                        metadatas=metadata_dict_final,
+                                        ids=str_uuid_file,
+                                    )
+                                    print("Added content.")
+                                    new_count +=1
```
Comment (markmcd) [inline demos/palm/python/docs-agent/scripts/populate_vector_database.py:277]: Could you run `black` over this? NBD but we can keep things consistent.

File: demos/palm/python/docs-agent/pylintrc [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+# Disable the message, report, category or checker with the given id(s). You
+# can either give multiple identifiers separated by comma (,) or put this
+# option multiple times (only on the command line, not in the configuration
+# file where it should appear only once).You can also use "--disable=all" to
+# disable everything first and then reenable specific checks. For example, if
+# you want to run only the similarities checker, you can use "--disable=all
+# --enable=similarities". If you want to run only the classes checker, but have
+# no Warning level messages displayed, use"--disable=all --enable=classes
+# --disable=W"
+disable=abstract-method,
```
Comment (markmcd) [inline demos/palm/python/docs-agent/pylintrc:53]: Why so much disabled? (apologies if this is a standard config, I've only ever used the defaults in google3)

**Review by markmcd — CHANGES_REQUESTED** [review]
Left a few basic comments. I didn't go over anything in detail, just spot checked a couple things and mainly checked the licenses.

Can you please include ownership changes in /CODEOWNERS and set up your unit test instrumentation too - can be another PR if you like, but please do it ASAP (e.g. no CODEOWNERS means we have to manually triage any PRs / issues that are filed).

File: demos/palm/python/docs-agent/third_party/css/chatbox.css [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright (c) 2023 by Landgreen (https://codepen.io/lilgreenland/pen/pyVvqB)
```
[the remark this answers is not in the document]
Comment (kyolee415) [reply demos/palm/python/docs-agent/third_party/css/chatbox.css:2]: It's a little bit involved.

After visiting the link (https://codepen.io/lilgreenland/pen/pyVvqB), you need to click **comments** on the bottom left corner of the window. Then the site opens a widget. At the bottom of the widget shows this exact license (which has been copied and pasted verbatim).

Please let me know if there is additional information (such as the instructions above) should be included here.

File: demos/palm/python/docs-agent/pylintrc (hunk 2 above)
Comment (rundong08) [reply demos/palm/python/docs-agent/pylintrc:53]: This is the official (open-source) google `pylintrc` obtained from https://google.github.io/styleguide/pylintrc.

File: demos/palm/python/docs-agent/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Contributing to Docs Agent
+
+Docs Agent lets anyone contribute to the project, regardless of their employer.
```
[the remark this answers is not in the document]
Comment (kyolee415) [reply demos/palm/python/docs-agent/CONTRIBUTING.md:3]: Will use the following `contributing.md` content instread:

https://github.com/google/new-project/blob/master/docs/contributing.md

File: demos/palm/python/docs-agent/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+For more information about the Google CLA,
+see [Contributor License Agreements](https://cla.developers.google.com/about).
+
+## Contributing changes and submitting code reviews
+
+All changes require review, including changes by project members.
+
+For detailed instructions on how to contribute changes,
+see README.md.
```
[the remark this answers is not in the document]
Comment (kyolee415) [reply demos/palm/python/docs-agent/CONTRIBUTING.md:23]: Same as above. Will use the following `contributing.md` content instead:

https://github.com/google/new-project/blob/master/docs/contributing.md

File: demos/palm/python/docs-agent/README.md [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+
+Keep in mind that this approach is not to “fine-tune” an LLM (large language model)
+itself, but the Docs Agent sample app uses a mixture of prompt engineering and
+embeddings techniques on top of a publicly available LLM model such as PaLM 2.
+
+![Docs Agent architecture](docs/images/docs-agent-architecture-01.png)
+
+**Figure 1**. Docs Agent uses a vector database to retrieve context for augmenting prompts.
+
+However, once you launch this chatbot using your own set of documents, you may soon realize
```
[the remark this answers is not in the document]
Comment (kyolee415) [reply demos/palm/python/docs-agent/README.md:32]: Good point! Will remove this paragraph from the introduction section.

File: demos/palm/python/docs-agent/scripts/populate_vector_database.py (hunk 1 above)
Comment (kyolee415) [reply demos/palm/python/docs-agent/scripts/populate_vector_database.py:277]: Sure! Will run `black` again.

File: demos/palm/python/docs-agent/CONTRIBUTING.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kyolee415) [reply demos/palm/python/docs-agent/CONTRIBUTING.md]: Updated the content of `CONTRIBUTING.md` to be https://github.com/google/new-project/blob/master/docs/contributing.md

File: demos/palm/python/docs-agent/CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kyolee415) [reply demos/palm/python/docs-agent/CONTRIBUTING.md]: Updated the content of `CONTRIBUTING.md` to be https://github.com/google/new-project/blob/master/docs/contributing.md

File: demos/palm/python/docs-agent/README.md [hunk 7]
```diff
@@ -0,0 +1,653 @@
+# Docs Agent
+
+The Docs Agent demo enables [PaLM API][genai-doc-site] users to launch a chat application
```
[the remark this answers is not in the document]
Comment (kyolee415) [reply demos/palm/python/docs-agent/README.md:3]: Done!

File: demos/palm/python/docs-agent/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (kyolee415) [reply demos/palm/python/docs-agent/README.md]: This paragraph has been removed from the `README.md` file.

File: demos/palm/python/docs-agent/scripts/populate_vector_database.py (hunk 1 above)
Comment (kyolee415) [reply demos/palm/python/docs-agent/scripts/populate_vector_database.py]: Updated!

File: CODEOWNERS [hunk 8]
```diff
@@ -6,5 +6,6 @@ demos/palm/web/travel-planner @lyleaf @jayjicheng
 demos/palm/web/list-it @mrayinteractive @aaron-wade
 demos/palm/web/quick-prompt @mrayinteractive @aaron-wade
 demos/palm/web/textfx @mrayinteractive @aaron-wade
+demos/palm/python/docs-agent @nickvander @rundong08 @Meggin @kyolee415
```
Comment (kyolee415) [inline CODEOWNERS:9]: I just included everyone who contributed to the Docs Agent project. Please take a look and let me know how to resolve the errors mentioned on this file. Thanks!

Comment (kyolee415) [reply CODEOWNERS:9]: Also yes, I'll submit updates on unit test setup in a follow-up pull request. Thanks!

Comment (markmcd) [reply CODEOWNERS:9]: Cool - these should all be fixed now, though you might each have to accept an invitation?

File: demos/palm/python/docs-agent/pylintrc (hunk 2 above)
Comment (markmcd) [reply demos/palm/python/docs-agent/pylintrc:53]: Oh crazy. Thanks for the info.

File: demos/palm/python/docs-agent/third_party/css/chatbox.css (hunk 3 above)
[the remark this answers is not in the document]
Comment (markmcd) [reply demos/palm/python/docs-agent/third_party/css/chatbox.css:2]: That's perfect - thank you!

**Review by markmcd — APPROVED** [review]
Thanks!

**Pull request merged**